[Core] Refactoring timer and adding timer_process

loumalouomega opened the pull request:
Fixes #5227. This PR refactors the timer, basically moves most of the code to the cpp and exposes more methods to python. And adds a process to generate the timing

loumalouomega committed:
Moving cpp to utilities folder (consistency)

loumalouomega committed:
Refactoring to cpp

loumalouomega committed:
Wrong description

loumalouomega committed:
Reverting (some compilation warnings)

loumalouomega committed:
Exposing stuff to python

loumalouomega committed:
Adding TimerProcess

loumalouomega committed:
Typos

loumalouomega committed:
Fix

loumalouomega committed:
More methods

loumalouomega committed:
Improving process

philbucher reviewed (commented):
Nice work!
We could take the to merge this timer with the „builtin_timer.h“ (which already uses the c++11 features std::chrono)

loumalouomega commented:
> Nice work!
> We could take the to merge this timer with the „builtin_timer.h“ (which already uses the c++11 features std::chrono)

Can we do that in a future PR?, otherwise maybe this one will be too much

loumalouomega commented:
@philbucher the current Timer does not depend on boost, but in standard c functions (not chrono). After checking it, I think it is better to proceed with this PR and improve the builtin_timer and integrate into timer. There are methods missing in order to be 100% compatible, and better to do one thing at a time.

philbucher commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,128 @@
+//
+
+// System includes
+#include <ctime>
```
I would replace this for std::chrono, then it is also independent of OMP
@msandre [did some tests](https://github.com/KratosMultiphysics/Kratos/pull/1478#issuecomment-366240165)

philbucher commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,128 @@
+{
+    if(msOutputFile.is_open())
+        PrintTimingInformation(msOutputFile);
+    else if(msPrintOnScreen)
```
why not possible to have both?

loumalouomega replied:
The original code is not mine

marcnunezc replied:
We could add the option to have both by removing the `else` no?

philbucher commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,128 @@
+#ifndef _OPENMP
+double Timer::msGlobalStart = std::clock()/static_cast<double>(CLOCKS_PER_SEC);
+#else
+double Timer::msGlobalStart = omp_get_wtime();
```
that stuff becomes much simpler when using chrono => check the [buitin_timer](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/builtin_timer.h)

loumalouomega replied:
OK

philbucher commented on kratos/utilities/timer.h:
```diff
@@ -20,14 +20,12 @@
-#include <ctime>
-
 #ifdef _OPENMP
 #include <omp.h>
```
can be removed when using chrono

philbucher commented on kratos/utilities/timer.h:
```diff
@@ -20,14 +20,12 @@
 
 // External includes
-#include <boost/timer.hpp> // to be removed after replacing the boost timers with Kratos timer.
+#include <boost/timer.hpp> // To be removed after replacing the boost timers with Kratos timer. Additionallty to be moved to the cpp
```
same

loumalouomega replied:
I cannot remove #include <boost/timer.hpp> because due to some lazy practices on Kratos code, people include the timer to later use the boost timer directly (instead of using the timer or importing the boost library directly). I  I remove it it breaks the compilation of some files. I propose to remove it later in a different PR

philbucher commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,128 @@
+Timer::ContainerType Timer::msTimeTable;
+std::ofstream Timer::msOutputFile;
+
+bool Timer::msPrintOnScreen = false;
```
why are those variables here and not in the header?

loumalouomega replied:
Because are static

philbucher commented on kratos/python_scripts/timer_process.py:
```diff
@@ -0,0 +1,65 @@
+    return TimerProcess(Model, settings["Parameters"])
+
+# All the processes python processes should be derived from "Process"
+class TimerProcess(KratosMultiphysics.Process):
```
@marcnunezc can you please try this?

philbucher reviewed (commented):
I am not sure how much you changed and how much was already there, many things show up as changed

loumalouomega committed:
Using std::chrono

loumalouomega commented:
@philbucher I used sd::chrono, the rest of your commentaries I replied directly

philbucher commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,123 @@
+//
+
+// System includes
+#include <ctime>
```
```suggestion

```

can be removed now

loumalouomega replied:
True

philbucher reviewed (commented):
good from my side now, approval I leave to @marcnunezc

loumalouomega committed:
Unused

marcnunezc commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,122 @@
+        KRATOS_INFO("Timer") << rIntervalName << " ";
+
+        for(int i = rIntervalName.size() + 1 ; i < 40 ; i++)
+            KRATOS_INFO("Timer") << ".";
```
```suggestion
            KRATOS_INFO("") << ".";
           
```

We need to remove the label here, otherwise this is the output:

`Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .Timer:  .`

Instead of

`...................`

marcnunezc commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,122 @@
+        for(int i = rIntervalName.size() + 1 ; i < 40 ; i++)
+            KRATOS_INFO("Timer") << ".";
+
+        KRATOS_INFO("Timer") << " " << StartTime << "s     \t" << StopTime << "s     \t" << StopTime - StartTime <<"s" << std::endl;
```
Here as well.

(line 84 is fine)

marcnunezc commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,122 @@
+            rOStream << ".";
+
+        rOStream << " ";
+        r_time_data.second.PrintData(rOStream, global_elapsed_time);
```
Can we add format here to `global_elapse_time`? So that all outputs have something like 5 decimals or so.

Different lengths of the timing screw up the format of the final table (see SolverInitialize):

```
Build.................................. 17 	0.227601s     	0.0146794s     	0.0121621s     	0.0133883s     	3.49373%
InitializeSolutionStep................. 5 	0.313942s     	0.0746375s     	0.0591535s     	0.0627885s     	4.81908%
MatrixStructure........................ 5 	0.146374s     	0.0393995s     	0.025908s     	0.0292748s     	2.24688%
Reading Input.......................... 1 	0.217838s     	0.217838s     	0.217838s     	0.217838s     	3.34387%
Solve.................................. 17 	2.73678s     	0.568448s     	0.117503s     	0.160987s     	42.0102%
SolveSolutionStep...................... 5 	3.05012s     	1.26766s     	0.435465s     	0.610024s     	46.82%
SolverInitialize....................... 1 	0.000392116s     	0.000392116s     	0.000392116s     	0.000392116s     	0.00601907%
Writing Mesh........................... 1 	0.0425325s     	0.0425325s     	0.0425325s     	0.0425325s     	0.652883%
Writing Results........................ 10 	0.0472715s     	0.00690678s     	0.0025723s     	0.00472715s     	0.725628%

```

marcnunezc replied:
Also, it seems the final timings are sorted alphabetically. Is it viable to sort them by order of call?

loumalouomega replied:
The structure is a map, not unordered map, the order is according to name

marcnunezc commented on kratos/utilities/timer.cpp:
```diff
@@ -0,0 +1,122 @@
+    msPrintOnScreen = PrintOnScreen;
+}
+
+void Timer::PrintIntervalInformation(std::string const& rIntervalName, const double StartTime, const double StopTime)
```
Right now each interval timing is always printed. Can we make this optional with a flag? That can be set by SetIntervalPrint() or similar.

I would say that most of the time you are interested in the accumulated total time of each process, that is printed in PrintTimingInformation().

marcnunezc reviewed (changes requested):
HI @loumalouomega, thanks for this!

I have been testing it and aside from the comments below, I have some questions:

- What is the expected usage of the python process? When testing it I have seen that it is too cumbersome to use, since you have to do something like:


    ```
    from KratosMultiphysics.timer_process import TimerProcess
    self.timer = TimerProcess(model, KratosMultiphysics.Parameters())
    self.InitializeSolutionStep()
    self.timer.ExecuteFinalize()
    ```

    Where I would have to define the proper paramaters with the interval name. If this is done once it is not too much of a problem, but if i wanted to add many timers, then I would need to create different parameters for each of those.

    I think that the python-interface should be as simple as if you were using `import time` and `time.time()` from python.

    If the scope of the python timer process is to be used once, then I would say that it is fine.

- However, if I directly use the CPP process as you do in the process, I like how it looks (it's even simpler than using time.time()!). Here is an example:

    First I define in my custom analysis stage the process with the options I need:
    ```
            self.timer = KratosMultiphysics.Timer()
            self.timer.SetPrintOnScreen(True)
            self.timer.SetOuputFile("timing.dat")
    ```

    And then in the overriden RunSolutionLoop:
    ```
    def RunSolutionLoop(self):
            while self.KeepAdvancingSolutionLoop():
                self.time = self._GetSolver().AdvanceInTime(self.time)
                self.timer.Start("InitializeSolutionStep")
                self.InitializeSolutionStep()
                self.timer.Stop("InitializeSolutionStep")

                self._GetSolver().Predict()

                self.timer.Start("SolveSolutionStep")
                is_converged = self._GetSolver().SolveSolutionStep()
                self.timer.Stop("SolveSolutionStep")

                self.FinalizeSolutionStep()
                self.OutputSolutionStep()
    ```

So I would personally use the CPP process rather than the python one.

loumalouomega committed:
Less verbosity

Co-Authored-By: Marc Núñez <mnunez@cimne.upc.edu>

loumalouomega committed:
Even less verbosity

loumalouomega committed:
Replacing the map for unordered_map

@pooyan-dadvand can you check that this can be done or do I revert. i do this in order to have the timings ordered in function of creation, not alphabetically

loumalouomega committed:
Removing boost::timer dependency

loumalouomega committed:
Both options

loumalouomega committed:
Adding msPrintIntervalInformation

loumalouomega committed:
Reducing code duplication

loumalouomega committed:
Better output

loumalouomega committed:
Improve process

loumalouomega committed:
Correct aspect

loumalouomega committed:
Consistent

loumalouomega committed:
Better format

loumalouomega committed:
Proper stream

loumalouomega committed:
Authorship

loumalouomega committed:
Final fix on spaces

loumalouomega commented:
@marcnunezc I adressed your comments

loumalouomega committed:
More methods

pooyan-dadvand commented:
My idea was to add the timing directly to the logger. Please give me some time to review this.

loumalouomega commented:
@marcnunezc we have green light from @pooyan-dadvand

marcnunezc commented:
Hi!
I tested it and I only have a few questios/ issues left. Here is one output I got after defining 4 "Custom" timer.Start() and timer.Stop()
```
                                 Repeat #		Total           	Max             	Min             	Average           	Time%
Custom4................................ 
Writing Results........................ 10		1.920244E-01s    	6.183561E-02s    	6.806087E-03s    	1.920244E-02s    	6.360641E+00%
Custom3................................ 1		1.925458E-01s    	1.925458E-01s    	1.925458E-01s    	1.925458E-01s    	6.377912E+00%
Solve.................................. 1		6.286523E-01s    	6.286523E-01s    	6.286523E-01s    	6.286523E-01s    	2.082356E+01%
Writing Mesh........................... 1		1.908270E-01s    	1.908270E-01s    	1.908270E-01s    	1.908270E-01s    	6.320976E+00%
MatrixStructure........................ 1		2.879648E-02s    	2.879648E-02s    	2.879648E-02s    	2.879648E-02s    	9.538580E-01%
Reading Input.......................... 1		7.973286E-01s    	7.973286E-01s    	7.973286E-01s    	7.973286E-01s    	2.641081E+01%
Custom1................................ 1		7.729597E-02s    	7.729597E-02s    	7.729597E-02s    	7.729597E-02s    	2.560361E+00%
Custom2................................ 1		6.444847E-01s    	6.444847E-01s    	6.444847E-01s    	6.444847E-01s    	2.134799E+01%
Build.................................. 1		1.262967E-02s    	1.262967E-02s    	1.262967E-02s    	1.262967E-02s    	4.183468E-01%
```

1) Seems like the order is still not by "appearence". See the code next. I dont need this to be in order 100% but I would like to understand why the final order is what it is in the end.

```
  def RunSolutionLoop(self):
		self.timer.Start("Custom1")
		self.InitializeSolutionStep()
		self.timer.Stop("Custom1")

		self._GetSolver().Predict()

		self.timer.Start("Custom2")
		is_converged = self._GetSolver().SolveSolutionStep()
		self.timer.Stop("Custom2")

		self.FinalizeSolutionStep()

		self.timer.Start("Custom3")
		self.OutputSolutionStep()
		self.timer.Stop("Custom3")
```

2) It seems i cannot time anything that its past the Finalize() call of the analysis stage. For instance doing:
      ```

	def Finalize(self):
		self.timer.Start("Custom4")
		super(PotentialAnalysisCustom,self).Finalize()
		self.timer.Stop("Custom4")

     ```
    Does not work as you can see in the table.

    Is that normal/inteded? I assume this is happening because we shutdown the timer at finalize or similiar.

3) There are other timers in the table that are there from other functions/processes in Kratos. Is it intended that those are there (Writing Mesh, Reading Input..)?

loumalouomega commented:
> Hi!
> I tested it and I only have a few questios/ issues left. Here is one output I got after defining 4 "Custom" timer.Start() and timer.Stop()
> 
> ```
>                                  Repeat #		Total           	Max             	Min             	Average           	Time%
> Custom4................................ 
> Writing Results........................ 10		1.920244E-01s    	6.183561E-02s    	6.806087E-03s    	1.920244E-02s    	6.360641E+00%
> Custom3................................ 1		1.925458E-01s    	1.925458E-01s    	1.925458E-01s    	1.925458E-01s    	6.377912E+00%
> Solve.................................. 1		6.286523E-01s    	6.286523E-01s    	6.286523E-01s    	6.286523E-01s    	2.082356E+01%
> Writing Mesh........................... 1		1.908270E-01s    	1.908270E-01s    	1.908270E-01s    	1.908270E-01s    	6.320976E+00%
> MatrixStructure........................ 1		2.879648E-02s    	2.879648E-02s    	2.879648E-02s    	2.879648E-02s    	9.538580E-01%
> Reading Input.......................... 1		7.973286E-01s    	7.973286E-01s    	7.973286E-01s    	7.973286E-01s    	2.641081E+01%
> Custom1................................ 1		7.729597E-02s    	7.729597E-02s    	7.729597E-02s    	7.729597E-02s    	2.560361E+00%
> Custom2................................ 1		6.444847E-01s    	6.444847E-01s    	6.444847E-01s    	6.444847E-01s    	2.134799E+01%
> Build.................................. 1		1.262967E-02s    	1.262967E-02s    	1.262967E-02s    	1.262967E-02s    	4.183468E-01%
> ```
> 
>     1. Seems like the order is still not by "appearence". See the code next. I dont need this to be in order 100% but I would like to understand why the final order is what it is in the end.

std::unordered_map is an implementation of hash table data structure, so it will arrange the elements internally according to the hash value using by std::unordered_map. But in case std::map it is usually a red black binary tree implementation. (Explanation)

> 
> 
> ```
>   def RunSolutionLoop(self):
> 		self.timer.Start("Custom1")
> 		self.InitializeSolutionStep()
> 		self.timer.Stop("Custom1")
> 
> 		self._GetSolver().Predict()
> 
> 		self.timer.Start("Custom2")
> 		is_converged = self._GetSolver().SolveSolutionStep()
> 		self.timer.Stop("Custom2")
> 
> 		self.FinalizeSolutionStep()
> 
> 		self.timer.Start("Custom3")
> 		self.OutputSolutionStep()
> 		self.timer.Stop("Custom3")
> ```
> 
>     1. It seems i cannot time anything that its past the Finalize() call of the analysis stage. For instance doing:
>        ```
>        
>        def Finalize(self):
>          self.timer.Start("Custom4")
>          super(PotentialAnalysisCustom,self).Finalize()
>          self.timer.Stop("Custom4")
>        ```
>        
>        
>        Does not work as you can see in the table.
>        Is that normal/inteded? I assume this is happening because we shutdown the timer at finalize or similiar.
> 

Are you using at the same time the timer_process, please don't, if you are planing to use in your own I recommend to create and close the timer by hand. (or maybe is related with the next question, see later)

>     2. There are other timers in the table that are there from other functions/processes in Kratos. Is it intended that those are there (Writing Mesh, Reading Input..)?

This is created from the model_part_io. Like erverything in timer is static it is like kind of global class

marcnunezc commented:
`Are you using at the same time the timer_process ...,`

Im directly using the CPP timer like this (I defined this in the init of my custom_analyis_stage):

```
		self.timer = KratosMultiphysics.Timer()
		self.timer.SetPrintOnScreen(True)
		self.timer.SetPrintIntervalInformation(False)
		self.timer.SetOuputFile("output.txt")

```

You dont recommend to do this?

loumalouomega commented:
> `Are you using at the same time the timer_process ...,`
> 
> Im directly using the CPP timer like this (I defined this in the init of my custom_analyis_stage):
> 
> ```
> 		self.timer = KratosMultiphysics.Timer()
> 		self.timer.SetPrintOnScreen(True)
> 		self.timer.SetPrintIntervalInformation(False)
> 		self.timer.SetOuputFile("output.txt")
> ```
> 
> You dont recommend to do this?

Maybe there is something closing the timer, as I said, everything is static, so it can be affected by other calls

marcnunezc commented:
Approving since the initial comments were addressed and the python process is functional.

Thanks @loumalouomega !

loumalouomega commented:
Thanks

pooyan-dadvand commented:
> @marcnunezc we have green light from @pooyan-dadvand

I cannot remember when I gave that green light. 

In fact I was in favour of putting the timer info inside the Logger and remove the Timer class.

loumalouomega commented:
I meant that the only significative we have change was the map->unordered_map. Actually in the end we reverted to map

loumalouomega commented:
We can move it to the logger, taking as base the current changes, which improved the previous version

RiccardoRossi commented:
guys, does this work correctly with openmp? (not all timers do)

philbucher commented:
> guys, does this work correctly with openmp? (not all timers do)

@RiccardoRossi see [my comment](https://github.com/KratosMultiphysics/Kratos/pull/5232#discussion_r303239703)

loumalouomega commented:
> > guys, does this work correctly with openmp? (not all timers do)
> 
> @RiccardoRossi see [my comment] [#5232 (comment)](https://github.com/KratosMultiphysics/Kratos/pull/5232#discussion_r303239703)

Thanks @philbucher
